z3rro/src/prover.rs

```diff
@@ -152,36 +243,76 @@ impl<'ctx> Prover<'ctx> {
         self.add_assumption(&value.not());
         self.min_level_with_provables.get_or_insert(self.level);
     }
-
+    
     /// `self.check_proof_assuming(&[])`.
-    pub fn check_proof(&mut self) -> ProveResult {
+    pub fn check_proof(&mut self) -> Result<ProveResult, ProverCommandError> {
+
         self.check_proof_assuming(&[])
     }
 
     /// Do the SAT check, but consider a check with no provables to be a
     /// [`ProveResult::Proof`].
-    pub fn check_proof_assuming(&mut self, assumptions: &[Bool<'ctx>]) -> ProveResult {
+    pub fn check_proof_assuming(
+        &mut self,
+        assumptions: &[Bool<'ctx>],
+    ) -> Result<ProveResult, ProverCommandError> {
         if !self.has_provables() {
-            return ProveResult::Proof;
+            return Ok(ProveResult::Proof);
         }
 
-        let res = match &self.last_result {
-            Some(cached_result) if assumptions.is_empty() => cached_result.last_result,
-            _ => {
-                let solver = self.get_solver();
-                let res = if assumptions.is_empty() {
-                    solver.check()
-                } else {
-                    solver.check_assumptions(assumptions)
+        match self.smt_solver {
+            SolverType::SWINE => {
+                let mut smtlib = self.get_smtlib();
+                smtlib.add_check_sat();
+                let smtlib = smtlib.into_string();
+                let mut smt_file: NamedTempFile = NamedTempFile::new().unwrap();
+                smt_file
+                    .write_all(remove_lines_for_swine(&smtlib).as_bytes())
+                    .unwrap();
+                let file_path = smt_file.path();
+
+                let res = execute_swine(file_path);
+                match res {
+                    Ok((SatResult::Unsat, _)) => Ok(ProveResult::Proof),
+                    Ok((SatResult::Unknown, reason)) => {
+                        Ok(ProveResult::Unknown(ReasonUnknown::Other(reason)))
+                    }
+                    Ok((SatResult::Sat, cex)) => {
+                        match &self.solver {
+                            StackSolver::Native(solver) => {
+                                solver.from_string(cex);
+                            },
+                            StackSolver::Emulated(solver, _) => {
+                                solver.from_string(cex);
+                            },
+                        }
+                        
+                        Ok(ProveResult::Counterexample)
+                    }
+                    Err(err) => Err(err),
+                }
+            }
+            SolverType::Z3 => {
+                let res = match &self.last_result {
```

Comment:
All of this logic needs to happen for the SWINE calls as well! This includes returning cached results, checking with assumptions (where did the assumptions go when using SWINE??) and caching results afterwards.